File: log-analyzer/src/services/__tests__/SearchQueryBuilder.test.ts
import { SearchQueryBuilder } from '../SearchQueryBuilder';

describe('SearchQueryBuilder', () => {
  describe('create', () => {
    it('should create empty query', () => {
      const builder = SearchQueryBuilder.create();
      const query = builder.getQuery();
      
      expect(query.terms).toHaveLength(0);
      expect(query.globalOperator).toBe('AND');
      expect(query.id).toBeTruthy();
      expect(query.metadata.createdAt).toBeGreaterThan(0);
    });
  });

  describe('fromString', () => {
    it('should parse single keyword', () => {
      const builder = SearchQueryBuilder.fromString('error');
      const query = builder.getQuery();
      
      expect(query.terms).toHaveLength(1);
      expect(query.terms[0].value).toBe('error');
      expect(query.terms[0].source).toBe('user');
    });

    it('should parse multiple keywords', () => {
      const builder = SearchQueryBuilder.fromString('error | timeout');
      const query = builder.getQuery();
      
      expect(query.terms).toHaveLength(2);
      expect(query.terms[0].value).toBe('error');
      expect(query.terms[1].value).toBe('timeout');
    });

    it('should handle empty string', () => {
      const builder = SearchQueryBuilder.fromString('');
      const query = builder.getQuery();
      
      expect(query.terms).toHaveLength(0);
    });

    it('should trim whitespace', () => {
      const builder = SearchQueryBuilder.fromString('  error  |  timeout  ');
      const query = builder.getQuery();
      
      expect(query.terms).toHaveLength(2);
      expect(query.terms[0].value).toBe('error');
      expect(query.terms[1].value).toBe('timeout');
    });

    it('should detect preset keywords', () => {
      const keywordGroups = [
        {
          id: 'group1',
          patterns: [{ regex: 'error' }]
        }
      ];
      
      const builder = SearchQueryBuilder.fromString('error', keywordGroups);
      const query = builder.getQuery();
      
      expect(query.terms[0].source).toBe('preset');
      expect(query.terms[0].presetGroupId).toBe('group1');
    });
  });

  describe('addTerm', () => {
    it('should add single term', () => {
      const builder = SearchQueryBuilder.create().addTerm('error');
      const query = builder.getQuery();
      
      expect(query.terms).toHaveLength(1);
      expect(query.terms[0].value).toBe('error');
      expect(query.terms[0].enabled).toBe(true);
      expect(query.terms[0].operator).toBe('AND');
    });

    it('should add multiple terms', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('timeout');
      
      const query = builder.getQuery();
      expect(query.terms).toHaveLength(2);
    });

    it('should respect custom options', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error', {
          source: 'preset',
          priority: 10,
          isRegex: true,
          operator: 'OR',
          caseSensitive: true
        });
      
      const query = builder.getQuery();
      expect(query.terms[0].source).toBe('preset');
      expect(query.terms[0].priority).toBe(10);
      expect(query.terms[0].isRegex).toBe(true);
      expect(query.terms[0].operator).toBe('OR');
      expect(query.terms[0].caseSensitive).toBe(true);
    });

    it('should update metadata on add', () => {
      const builder = SearchQueryBuilder.create();
      const beforeTime = Date.now();
      
      builder.addTerm('error');
      
      const query = builder.getQuery();
      expect(query.metadata.lastModified).toBeGreaterThanOrEqual(beforeTime);
    });
  });

  describe('removeTerm', () => {
    it('should remove term by id', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('timeout');
      
      const termId = builder.getQuery().terms[0].id;
      builder.removeTerm(termId);
      
      expect(builder.getQuery().terms).toHaveLength(1);
      expect(builder.getQuery().terms[0].value).toBe('timeout');
    });

    it('should handle non-existent id', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error');
      
      builder.removeTerm('non-existent');
      
      expect(builder.getQuery().terms).toHaveLength(1);
    });
  });

  describe('toggleTerm', () => {
    it('should toggle term enabled state', () => {
      const builder = SearchQueryBuilder.create().addTerm('error');
      const termId = builder.getQuery().terms[0].id;
      
      expect(builder.getQuery().terms[0].enabled).toBe(true);
      
      builder.toggleTerm(termId);
      expect(builder.getQuery().terms[0].enabled).toBe(false);
      
      builder.toggleTerm(termId);
      expect(builder.getQuery().terms[0].enabled).toBe(true);
    });

    it('should handle non-existent id', () => {
      const builder = SearchQueryBuilder.create().addTerm('error');
      
      expect(() => builder.toggleTerm('non-existent')).not.toThrow();
    });
  });

  describe('updateTermValue', () => {
    it('should update term value', () => {
      const builder = SearchQueryBuilder.create().addTerm('error');
      const termId = builder.getQuery().terms[0].id;
      
      builder.updateTermValue(termId, 'warning');
      
      expect(builder.getQuery().terms[0].value).toBe('warning');
    });
  });

  describe('clear', () => {
    it('should remove all terms', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('timeout')
        .clear();
      
      expect(builder.getQuery().terms).toHaveLength(0);
    });
  });

  describe('validate', () => {
    it('should fail on empty query', () => {
      const builder = SearchQueryBuilder.create();
      const result = builder.validate();
      
      expect(result.isValid).toBe(false);
      expect(result.issues[0].code).toBe('EMPTY_QUERY');
    });

    it('should pass on valid query', () => {
      const builder = SearchQueryBuilder.create().addTerm('error');
      const result = builder.validate();
      
      expect(result.isValid).toBe(true);
      expect(result.issues).toHaveLength(0);
    });

    it('should detect empty term value', () => {
      const builder = SearchQueryBuilder.create().addTerm('  ');
      const result = builder.validate();
      
      expect(result.isValid).toBe(false);
      expect(result.issues.some(i => i.code === 'EMPTY_VALUE')).toBe(true);
    });

    it('should detect invalid regex', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('[invalid', { isRegex: true });
      const result = builder.validate();
      
      expect(result.isValid).toBe(false);
      expect(result.issues.some(i => i.code === 'INVALID_REGEX')).toBe(true);
    });

    it('should warn on duplicate terms', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('error');
      const result = builder.validate();
      
      const warnings = result.issues.filter(i => i.severity === 'warning');
      expect(warnings.length).toBeGreaterThan(0);
      expect(warnings.some(w => w.code === 'DUPLICATE_TERM')).toBe(true);
    });

    it('should detect disabled terms', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error');
      const termId = builder.getQuery().terms[0].id;
      builder.toggleTerm(termId);
      
      const result = builder.validate();
      expect(result.isValid).toBe(false);
      expect(result.issues[0].code).toBe('NO_ENABLED_TERMS');
    });

    it('should warn on value too long', () => {
      const longValue = 'a'.repeat(101);
      const builder = SearchQueryBuilder.create().addTerm(longValue);
      const result = builder.validate();
      
      const warnings = result.issues.filter(i => i.severity === 'warning');
      expect(warnings.some(w => w.code === 'VALUE_TOO_LONG')).toBe(true);
    });
  });

  describe('toQueryString', () => {
    it('should generate query string', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('timeout');
      
      const queryString = builder.toQueryString();
      expect(queryString).toBe('error | timeout');
    });

    it('should exclude disabled terms', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('timeout');
      
      const termId = builder.getQuery().terms[0].id;
      builder.toggleTerm(termId);
      
      const queryString = builder.toQueryString();
      expect(queryString).toBe('timeout');
    });

    it('should return empty string for empty query', () => {
      const builder = SearchQueryBuilder.create();
      expect(builder.toQueryString()).toBe('');
    });
  });

  describe('toOptimizedQuery', () => {
    it('should create optimized query', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('timeout');
      
      const optimized = builder.toOptimizedQuery();
      expect(optimized.enabledTerms).toHaveLength(2);
      expect(optimized.termsByOperator.get('AND')).toHaveLength(2);
    });

    it('should group by operator', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error', { operator: 'AND' })
        .addTerm('warning', { operator: 'OR' })
        .addTerm('debug', { operator: 'NOT' });
      
      const optimized = builder.toOptimizedQuery();
      expect(optimized.termsByOperator.get('AND')).toHaveLength(1);
      expect(optimized.termsByOperator.get('OR')).toHaveLength(1);
      expect(optimized.termsByOperator.get('NOT')).toHaveLength(1);
    });

    it('should calculate normalization stats', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('error') // duplicate
        .addTerm('timeout');
      
      const termId = builder.getQuery().terms[2].id;
      builder.toggleTerm(termId); // disable one
      
      const optimized = builder.toOptimizedQuery();
      expect(optimized.normalization.removedInvalidCount).toBe(1);
    });
  });

  describe('export and import', () => {
    it('should export and import query', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('timeout');
      
      const exported = builder.export();
      const imported = SearchQueryBuilder.import(exported);
      
      expect(imported.getQuery().terms).toHaveLength(2);
      expect(imported.toQueryString()).toBe(builder.toQueryString());
    });

    it('should preserve all properties', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error', {
          source: 'preset',
          presetGroupId: 'group1',
          isRegex: true,
          priority: 10
        });
      
      const exported = builder.export();
      const imported = SearchQueryBuilder.import(exported);
      const term = imported.getQuery().terms[0];
      
      expect(term.source).toBe('preset');
      expect(term.presetGroupId).toBe('group1');
      expect(term.isRegex).toBe(true);
      expect(term.priority).toBe(10);
    });
  });

  describe('utility methods', () => {
    it('should set global operator', () => {
      const builder = SearchQueryBuilder.create()
        .setGlobalOperator('OR');
      
      expect(builder.getQuery().globalOperator).toBe('OR');
    });

    it('should get summary', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error', { source: 'preset', isRegex: true })
        .addTerm('timeout');
      
      const summary = builder.getSummary();
      expect(summary.termCount).toBe(2);
      expect(summary.enabledTermCount).toBe(2);
      expect(summary.hasPreset).toBe(true);
      expect(summary.hasRegex).toBe(true);
    });

    it('should check if has value', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error');
      
      expect(builder.has('error')).toBe(true);
      expect(builder.has('ERROR')).toBe(true); // case insensitive
      expect(builder.has('timeout')).toBe(false);
    });

    it('should find term by value', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error');
      
      const term = builder.findTermByValue('error');
      expect(term).toBeDefined();
      expect(term?.value).toBe('error');
      
      const notFound = builder.findTermByValue('timeout');
      expect(notFound).toBeUndefined();
    });

    it('should get all terms', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('timeout');
      
      const terms = builder.getTerms();
      expect(terms).toHaveLength(2);
    });

    it('should get enabled terms only', () => {
      const builder = SearchQueryBuilder.create()
        .addTerm('error')
        .addTerm('timeout');
      
      const termId = builder.getQuery().terms[0].id;
      builder.toggleTerm(termId);
      
      const enabledTerms = builder.getEnabledTerms();
      expect(enabledTerms).toHaveLength(1);
      expect(enabledTerms[0].value).toBe('timeout');
    });
  });

  describe('toJSON', () => {
    it('should return deep copy', () => {
      const builder = SearchQueryBuilder.create().addTerm('error');
      const json = builder.toJSON();
      
      // Modify the copy
      json.terms[0].value = 'modified';
      
      // Original should be unchanged
      expect(builder.getQuery().terms[0].value).toBe('error');
    });
  });
});

File: log-analyzer_flutter/test/shared/services/search_query_builder_test.dart
// SearchQueryBuilder 测试
//
// 遵循编码规范：使用 flutter_test

import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:log_analyzer_flutter/shared/services/search_query_builder.dart';
import 'package:log_analyzer_flutter/core/constants/app_constants.dart';

void main() {
  group('SearchQueryBuilder', () {
    group('create', () {
      test('应该创建空的查询构建器', () {
        final builder = SearchQueryBuilder.create();

        expect(builder.termCount, 0);
        expect(builder.enabledTermCount, 0);
        expect(builder.id, isNotEmpty);
      });
    });

    group('fromString', () {
      test('应该解析简单关键词', () {
        final builder = SearchQueryBuilder.fromString('error');
        final query = builder.toQuery();

        expect(query.terms, hasLength(1));
        expect(query.terms.first.value, 'error');
        expect(query.terms.first.operator_.value, 'OR');
      });

      test('应该解析多个关键词（OR 模式）', () {
        final builder = SearchQueryBuilder.fromString(
          'error | warning | fatal',
        );
        final query = builder.toQuery();

        expect(query.terms, hasLength(3));
        expect(query.terms[0].value, 'error');
        expect(query.terms[1].value, 'warning');
        expect(query.terms[2].value, 'fatal');
      });

      test('应该处理空输入', () {
        final builder = SearchQueryBuilder.fromString('');
        final query = builder.toQuery();

        expect(query.terms, isEmpty);
      });

      test('应该去除多余空格', () {
        final builder = SearchQueryBuilder.fromString(
          '  error   |   warning  ',
        );
        final query = builder.toQuery();

        expect(query.terms, hasLength(2));
      });
    });

    group('addTerm', () {
      test('应该添加单个搜索术语', () {
        final builder = SearchQueryBuilder.create().addTerm('error');

        expect(builder.termCount, 1);
        expect(builder.enabledTermCount, 1);
      });

      test('应该添加多个搜索术语', () {
        final builder = SearchQueryBuilder.create()
            .addTerm('error')
            .addTerm('warning')
            .addTerm('fatal');

        expect(builder.termCount, 3);
      });

      test('应该支持正则表达式', () {
        final builder = SearchQueryBuilder.create().addTerm(
          r'\berror\d+\b',
          isRegex: true,
        );
        final query = builder.toQuery();

        expect(query.terms.first.isRegex, true);
        expect(query.terms.first.value, r'\berror\d+\b');
      });

      test('应该支持大小写敏感', () {
        final builder = SearchQueryBuilder.create().addTerm(
          'Error',
          caseSensitive: true,
        );
        final query = builder.toQuery();

        expect(query.terms.first.caseSensitive, true);
      });

      test('应该支持优先级设置', () {
        final builder = SearchQueryBuilder.create().addTerm(
          'fatal',
          priority: 10,
        );
        final query = builder.toQuery();

        expect(query.terms.first.priority, 10);
      });
    });

    group('removeTerm', () {
      test('应该移除指定术语', () {
        final builder = SearchQueryBuilder.create()
            .addTerm('error')
            .addTerm('warning');
        final termId = builder.toQuery().terms.first.id;

        final updated = builder.removeTerm(termId);

        expect(updated.termCount, 1);
        expect(updated.toQuery().terms.first.value, 'warning');
      });

      test('移除不存在的术语应该无效果', () {
        final builder = SearchQueryBuilder.create().addTerm('error');

        final updated = builder.removeTerm('non-existent-id');

        expect(updated.termCount, 1);
      });
    });

    group('toggleTerm', () {
      test('应该切换术语启用状态', () {
        final builder = SearchQueryBuilder.create().addTerm('error');
        final termId = builder.toQuery().terms.first.id;

        final disabled = builder.toggleTerm(termId);
        expect(disabled.enabledTermCount, 0);

        final enabled = disabled.toggleTerm(termId);
        expect(enabled.enabledTermCount, 1);
      });
    });

    group('clear', () {
      test('应该清空所有术语', () {
        final builder = SearchQueryBuilder.create()
            .addTerm('error')
            .addTerm('warning');

        final cleared = builder.clear();

        expect(cleared.termCount, 0);
      });
    });

    group('setGlobalOperator', () {
      test('应该设置全局操作符', () {
        final builder = SearchQueryBuilder.create()
            .addTerm('error')
            .setGlobalOperator(QueryOperator.and);

        final query = builder.toQuery();
        expect(query.globalOperator.value, 'AND');
      });
    });

    group('toQueryString', () {
      test('应该生成 OR 模式查询字符串', () {
        final queryString = SearchQueryBuilder.create()
            .addTerm('error')
            .addTerm('warning')
            .toQueryString();

        expect(queryString, contains('error'));
        expect(queryString, contains('|'));
        expect(queryString, contains('warning'));
      });

      test('应该生成 AND 模式查询字符串', () {
        final queryString = SearchQueryBuilder.create()
            .addTerm('error')
            .addTerm('warning')
            .setGlobalOperator(QueryOperator.and)
            .toQueryString();

        expect(queryString, contains('error'));
        expect(queryString, contains('warning'));
        expect(queryString.contains('|'), false); // AND 模式使用空格分隔
      });

      test('应该排除禁用的关键词', () {
        var builder = SearchQueryBuilder.create()
            .addTerm('error')
            .addTerm('warning');
        final disabledTermId = builder.toQuery().terms.last.id;
        builder = builder.toggleTerm(disabledTermId);

        final queryString = builder.toQueryString();

        expect(queryString, contains('error'));
        expect(queryString, isNot(contains('warning')));
      });
    });

    group('toOptimizedQuery', () {
      test('应该生成优化后的查询', () {
        final builder = SearchQueryBuilder.create()
            .addTerm('error', priority: 1)
            .addTerm('fatal', priority: 10);

        final optimized = builder.toOptimizedQuery();

        expect(optimized.queryString, isNotEmpty);
        expect(optimized.prioritizedTerms, hasLength(2));
        // 高优先级应该在前面
        expect(optimized.prioritizedTerms.first, 'fatal');
      });
    });

    group('validate', () {
      test('应该验证有效的查询', () {
        final builder = SearchQueryBuilder.create().addTerm('error');
        final result = builder.validate();

        expect(result.valid, true);
        expect(result.errors, isNull);
      });

      test('应该检测空查询', () {
        final builder = SearchQueryBuilder.create();
        final result = builder.validate();

        expect(result.valid, false);
        expect(result.errors, isNotEmpty);
      });

      test('应该检测没有启用的术语', () {
        var builder = SearchQueryBuilder.create().addTerm('error');
        final termId = builder.toQuery().terms.first.id;
        builder = builder.toggleTerm(termId);

        final result = builder.validate();

        expect(result.warnings, isNotEmpty);
      });

      test('应该检测无效的正则表达式', () {
        final builder = SearchQueryBuilder.create().addTerm(
          '[invalid(',
          isRegex: true,
        );
        final result = builder.validate();

        expect(result.valid, false);
        expect(result.errors, isNotEmpty);
      });
    });

    group('JSON 序列化', () {
      test('export 应该生成有效的 JSON', () {
        final builder = SearchQueryBuilder.create()
            .addTerm('error')
            .addTerm('warning');

        final json = builder.export();

        expect(json, isNotEmpty);
        // 应该是有效的 JSON
        expect(() => jsonDecode(json), returnsNormally);
      });

      test('import 应该从 JSON 恢复查询', () {
        final original = SearchQueryBuilder.create()
            .addTerm('error')
            .addTerm('warning')
            .setGlobalOperator(QueryOperator.and);

        final json = original.export();
        final restored = SearchQueryBuilder.import(json);
        final query = restored.toQuery();

        expect(query.terms, hasLength(2));
        expect(query.terms[0].value, 'error');
        expect(query.terms[1].value, 'warning');
        expect(query.globalOperator.value, 'AND');
      });

      test('import 无效 JSON 应该返回空构建器', () {
        final restored = SearchQueryBuilder.import('invalid json');

        expect(restored.termCount, 0);
      });
    });
  });
}
